tool/tsh/kubectl.go

```diff
@@ -0,0 +1,256 @@
+				// In the second case, we should create a Resource Access Request to allow
+				// the user to exec/read logs using different Kubernetes RBAC principals.
+				// using different Kubernetes RBAC principals.
+				if podForbiddenRe.MatchString(line) {
+					results := podForbiddenRe.FindStringSubmatch(line)
+					missingKubeResources = append(missingKubeResources, resourceKind{kind: types.KindKubePod, subResourceName: filepath.Join(results[2], results[1])})
+					// Check if cluster access was denied. If denied we should create
+					// a Resource Access Request for the cluster and not a pod.
+				} else if strings.Contains(line, clusterForbidden) {
+					missingKubeResources = append(missingKubeResources, resourceKind{kind: types.KindKubernetesCluster})
+				}
+			}
+		}()
+
+		err := runKubectlReexec(cf.executablePath, args, writer)
+		if err == nil {
+			break
+		} else if !errors.As(err, &exitErr) {
+			return trace.Wrap(err)
+		} else if errors.As(err, &exitErr) && exitErr.ExitCode() != cmdutil.DefaultErrorExitCode {
+			// if the exit code is not 1, it was emitted by pod exec code and we should
+			// ignore it since the user was allowed to execute the command in the pod.
+			break
+		}
+		writer.CloseWithError(io.EOF)
+		wg.Wait()
+		if len(missingKubeResources) > 0 && !alreadyRequestedAccess {
+			// create the access requests for the user and wait for approval.
+			if err := createKubeAccessRequest(cf, missingKubeResources, args); err != nil {
+				return trace.Wrap(err)
+			}
+			alreadyRequestedAccess = true
+			continue
+		}
+		break
+	}
+	// exit with the kubectl exit code to keep compatibility.
+	if errors.As(err, &exitErr) {
+		os.Exit(exitErr.ExitCode())
+	}
+	return nil
+}
+
+// createKubeAccessRequest creates an access request to the denied resources
+// if the user's roles allow search_as_role.
+func createKubeAccessRequest(cf *CLIConf, resources []resourceKind, args []string) error {
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	kubeName, err := getKubeClusterName(args, tc.SiteName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, rec := range resources {
+		cf.RequestedResourceIDs = append(
+			cf.RequestedResourceIDs,
+			filepath.Join("/", tc.SiteName, rec.kind, kubeName, rec.subResourceName),
+		)
+	}
+	cf.Reason = fmt.Sprintf("Resource request automatically created for %v", args)
+	if err := executeAccessRequest(cf, tc); err != nil {
+		// TODO: intercept the error to validate the origin
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// extractKubeConfigAndContext parses the args and extracts - if present -
+// the --kubeconfig flag that overrrides the default kubeconfig location
+// and the --context flag that overrides the default context to use.
+func extractKubeConfigAndContext(args []string) (kubeconfig string, context string) {
+	// ignore any error on parse since we are only interested in two flag values.
+	fs := pflag.NewFlagSet("kubectl", pflag.ContinueOnError)
```

Comment:
should `"kubectl"` be a constant instead of hardcoded string?

Reply:
This string can be anything and it's only important if someone called `--help`.
Since this function is a helper function to extract the flag's value and it's never executed if we call `--help`, it's not important.